### PR TITLE
Create CKAN graph content type

### PR DESCRIPTION
**Pull request by lem-doghouse**

### CODE REVIEW

File: govcms_ckan.admin.inc [hunk 1]
```diff
@@ -56,4 +56,4 @@ function govcms_ckan_settings_form_validate($form, &$form_state) {
   if (!isset($data->success)) {
     form_set_error('govcms_ckan_endpoint_url', t('Could not establish a connection to the endpoint.'));
   }
-}
+}
```
Comment (jorgegc-doghouse) [inline govcms_ckan.admin.inc]: We need to have an empty line at the end of the file @lem-doghouse.

File: govcms_ckan.info [hunk 2]
```diff
@@ -2,4 +2,4 @@ name = govCMS CKAN
 description = govCMS CKAN provides integration with CKAN.
 package = govCMS
 core = 7.x
-configure = admin/config/services/govcms-ckan
+configure = admin/config/services/govcms-ckan
```
Comment (jorgegc-doghouse) [inline govcms_ckan.info]: We need this blank line at the end of the file. Can you please revert this change?

File: govcms_ckan.install [hunk 3]
```diff
@@ -6,9 +6,158 @@
  */
 
 /**
+ * Implements hook_field_schema().
+ */
+function govcms_ckan_field_schema($field) {
+  return array(
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: It is probably a good idea to check for the field type before returning the schema, specially when the module defines multiple field types.

```
$schema = array();
if ($field['type'] == 'my_custom_field_type') {
  ...
}
return $schema;
```

File: govcms_ckan.install [hunk 4]
```diff
@@ -6,9 +6,158 @@
...
+    'description' => st('Create a new graph that can be attach to a node or wysiwyg.'),
+    'custom' => TRUE,
+    'modified' => TRUE,
+    'locked' => FALSE,
+  );
+
+  $graph_type = node_type_set_defaults($graph_type);
+  node_type_save($graph_type);
+
+  // We enable the CKAN graph by default
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: This code doesn't comply with Drupal's coding standards because it is missing a full stop.

```
// Enable the CKAN graph functionality by default.
```

File: govcms_ckan.install [hunk 5]
```diff
@@ -6,9 +6,158 @@
...
+  variable_set('ckan_graph_node_ckan_graph', TRUE);
+
+  govcms_ckan_create_graph_data_field();
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: I think we should make these functions a little bit more flexible @lem-doghouse. How about we add two parameters to them (`$entity_type` and `$bundle`) to allow for these fields to be attached to any entity and bundle?

File: govcms_ckan.install [hunk 6]
```diff
@@ -6,9 +6,158 @@
...
+}
+
+/**
+ * Create the graph data field and binds it to ckan_graph bundle.
+ */
+function govcms_ckan_create_graph_data_field() {
+  $field_name = 'ckan_graph_data';
+  $field = array(
+    'field_name' => $field_name,
+    'type' => 'ckan_graph_field_data',
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: Can we please change this to `ckan_dataset`?

File: govcms_ckan.install [hunk 7]
```diff
@@ -6,9 +6,158 @@
...
+    'cardinality' => '-1',
+  );
+  $field = field_create_field($field);
+
+  // Creates the field instance and bind it to ckan_graph bundle.
+  $instance = array(
+    'field_name' => $field['field_name'],
+    'entity_type' => 'node',
+    'bundle' => 'ckan_graph',
+    'description' => 'A field for for storing CKAN graph dataset',
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: There is a typo in the description `for for`.

File: govcms_ckan.install [hunk 8]
```diff
@@ -6,9 +6,158 @@
...
+      'type' => 'ckan_graph_field_data',
+    ),
+  );
+  field_create_instance($instance);
+}
+
+/**
+ * Create the graph type field and binds it to ckan_graph bundle.
+ */
+function govcms_ckan_create_graph_type_field() {
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: As per my comment on line 58, we should probably add `$entity_type` and `$bundle` as parameters of this function to transform it more into an API-like function rather than a function that can only attach a particular field to a particular bundle of a particular entity type.

File: govcms_ckan.install [hunk 9]
```diff
@@ -6,9 +6,158 @@
...
+  $field_name = 'ckan_graph_type';
+
+  // We list the default supported graph types.
+  $graph_type = array(
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: This will most likely be moved to the custom graph type field widget.

File: govcms_ckan.install [hunk 10]
```diff
@@ -6,9 +6,158 @@
...
+
+  $field = array(
+    'field_name' => $field_name,
+    'type' => 'list_text',
+    'settings' => array(
+      'allowed_values' => $graph_type,
+      'allowed_values_function' => '',
+    ),
+  );
+  $field = field_create_field($field);
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: Can we please check whether the field already exists before attempting to create it?

File: govcms_ckan.install [hunk 11]
```diff
@@ -6,9 +6,158 @@
...
+    'description' => '',
+    'default_value' => NULL,
+    'widget' => array(
+      'weight' => '1',
+      'type' => 'options_select',
+      'module' => 'options',
+      'active' => 1,
+    ),
+  );
+  field_create_instance($instance);
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: Can we also please check whether the instance already exists before creating it?

File: govcms_ckan.install [hunk 12]
```diff
@@ -6,9 +6,158 @@
...
+    'custom' => TRUE,
+    'modified' => TRUE,
+    'locked' => FALSE,
+  );
+
+  $graph_type = node_type_set_defaults($graph_type);
+  node_type_save($graph_type);
+
+  // We enable the CKAN graph by default
+  variable_set('ckan_graph_node_ckan_graph', TRUE);
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: Because this is a variable @lem-doghouse, it is probably good to namespace it. Can we change it to `govcms_ckan_graph_node_ckan_graph`?

File: govcms_ckan.install [hunk 13]
```diff
@@ -6,9 +6,158 @@
...
+
+/**
  * Implements hook_uninstall().
  */
 function govcms_ckan_uninstall() {
   variable_del('govcms_ckan_endpoint_url');
   variable_del('govcms_ckan_api_key');
+
+  foreach (node_type_get_names() as $type => $name) {
+    variable_del('ckan_graph_node_' . $type);
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: Please change this variable to `'govcms_ckan_graph_node_' . $type` as per my previous comment.

File: govcms_ckan.module [hunk 14]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+function govcms_ckan_form_node_type_form_alter(&$form, $form_state) {
+  if (isset($form['type'])) {
+    $form['ckan_graph'] = array(
+      '#title' => t('CKAN Graph'),
+      '#type' => 'fieldset',
+      '#collapsible' => TRUE,
+      '#group' => 'additional_settings',
+      '#weight' => -100,
+    );
+    $form['ckan_graph']['ckan_graph_node'] = array(
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Can we change this to `$form['ckan_graph']['govcms_ckan_graph_node'] = array(`?

File: govcms_ckan.module [hunk 15]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+      '#type' => 'checkbox',
+      '#title' => t('Enable govCMS CKAN graph.'),
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Please update title to `Enable govCMS CKAN graph functionality.`

File: govcms_ckan.module [hunk 16]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+      '#description' => t('Allows govCMS CKAN Graph to be attached to this content type.'),
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Please change the description to `Allows this content type to be displayed as a CKAN graph.`

File: govcms_ckan.module [hunk 17]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+    );
+  }
+}
+
+/**
+ * Implements hook_field_info().
+ */
+function govcms_ckan_field_info() {
+  return array(
+    'ckan_graph_field_data' => array(
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: As per my previous comment, can we please change this to `ckan_dataset`?

File: govcms_ckan.module [hunk 18]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+      'label' => t('CKAN Graph Dataset'),
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: I think we should also update the label `CKAN Dataset`.

File: govcms_ckan.module [hunk 19]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+      'description' => t("This field stores CKAN graph data."),
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: For consistency, please only use single quotes. Would be good to change the description to `This field stores a reference to a CKAN dataset.`. The graph is only the visual representation of the dataset @lem-doghouse.

File: govcms_ckan.module [hunk 20]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+      'settings' => array(
+        'allowed_values' => array(),
+        'allowed_values_function' => ''
+      ),
+      'default_widget' => 'ckan_graph_field_data',
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Please namespace with the module name `govcms_ckan_graph_field_data`.

File: govcms_ckan.module [hunk 21]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+      'default_formatter' => 'ckan_graph_field_data_default',
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Please also namespace with the module name `govcms_ckan_graph_field_data_default `.

File: govcms_ckan.module [hunk 22]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+    ),
+  );
+}
+
+/**
+ * Implements hook_field_is_empty().
+ */
+function govcms_ckan_field_is_empty($item, $field) {
+  if ($field['type'] == 'ckan_graph_field_data') {
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: You are going to have to update this field type `ckan_dataset`.

File: govcms_ckan.module [hunk 23]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+    return empty($item['dataset_id']) || empty($item['dataset_title']);
+  }
+}
+
+/**
+ * Implements hook_field_widget_info().
+ */
+function govcms_ckan_field_widget_info() {
+  return array(
+    'ckan_graph_field_data' => array(
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: You are going to need a `switch` statement here to be able to handle multiple custom field widgets.

File: govcms_ckan.module [hunk 24]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+      'label' => t('CKAN Graph'),
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Because there is only one widget for this field type, I reckon it is safe to call it `CKAN Dataset`.

File: govcms_ckan.install [hunk 25]
```diff
@@ -6,9 +6,158 @@
...
+  $instance = array(
+    'field_name' => $field['field_name'],
+    'entity_type' => 'node',
+    'bundle' => 'ckan_graph',
+    'description' => 'A field for for storing CKAN graph dataset',
+    'label' => 'Data',
+    'required' => 1,
+    'active' => 1,
+    'widget' => array(
+      'type' => 'ckan_graph_field_data',
```
Comment (jorgegc-doghouse) [inline govcms_ckan.install]: Please change to `ckan_dataset`.

File: govcms_ckan.module (hunk 20 above)
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Please change to `ckan_dataset`.

File: govcms_ckan.module (hunk 21 above)
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Please change to `ckan_dataset_default`.

File: govcms_ckan.module [hunk 26]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+      'field types' => array('ckan_graph_field_data'),
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: You are going to need to update this to `ckan_dataset` @lem-doghouse.

File: govcms_ckan.module [hunk 27]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+    ),
+  );
+}
+
+/**
+ * Implements hook_field_widget_form().
+ */
+function govcms_ckan_field_widget_form(&$form, &$form_state, $field, $instance, $langcode, $items, $delta, $element) {
+  // We make sure that we're dealing with ckan graph field.
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Because we are going to call the field `ckan_dataset`, it makes sense to reword this to `Make sure that we are dealing with CKAN dataset field.` I think we are going to need a `switch` statement here too.

File: govcms_ckan.module [hunk 28]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+  if ($instance['widget']['type'] == 'ckan_graph_field_data') {
+    $fields = array(
+      'dataset_id' => t('Dataset id'),
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Can we please capitalise `ID`? Thanks

File: govcms_ckan.module [hunk 29]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+  }
+  return $element;
+}
+
+/**
+ * Implements hook_field_formatter_info().
+ */
+function govcms_ckan_field_formatter_info() {
+  return array(
+    'ckan_graph_field_data_default' => array(
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Can we change this to `ckan_dataset_default`?

File: govcms_ckan.module [hunk 30]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+      'label' => t('Default'),
+      'field types' => array('ckan_graph_field_data'),
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: The machine name of the field type should be `can_dataset`. Can we please update?

File: govcms_ckan.module [hunk 31]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+}
+
+/**
+ * Implements hook_field_formatter_view().
+ */
+function govcms_ckan_field_formatter_view($entity_type, $entity, $field, $instance, $langcode, $items, $display) {
+  $element = array();
+  $node = menu_get_object();
+
+  if (!empty($node)) {
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Any particular reason why we need to check whether we are looking at a `node` here?

File: govcms_ckan.module [hunk 32]
```diff
@@ -36,3 +36,116 @@ function govcms_ckan_permission() {
...
+    if ($display['type'] == 'ckan_graph_field_data_default') {
```
Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Please update to `ckan_dataset_default`. Also, this might need to be converted into a `switch` statement.

File: govcms_ckan.module (hunk 23 above)
Comment (lem-doghouse) [inline govcms_ckan.module]: Hi @jorgegc-doghouse,

Just confirming that I don't need a `switch` statement in hook_field_widget_info() as opposed to the suggestion above?

Comment (jorgegc-doghouse) [inline govcms_ckan.module]: Yes, you are right. There is no need for a `switch` statement here. My bad!
